plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Stablecoin.hs

```diff
@@ -301,13 +327,36 @@ stableCoins sc@Stablecoin{scStablecoinTokenName} =
 
 {-# INLINEABLE isValidState #-}
 isValidState :: Stablecoin -> BankState -> ConversionRate -> Bool
-isValidState sc bs@BankState{bsReservecoins, bsReserves, bsStablecoins} cr =
-    traceIfFalse "reservecoins < 0" (bsReservecoins >= RC 0)
-    && traceIfFalse "reserves < 0" (bsReserves >= BC 0)
-    && traceIfFalse "stablecoins < 0" (bsStablecoins >= SC 0)
-    && traceIfFalse "min reserves" (fmap fromInteger bsReserves >= minReserve sc cr bs)
-    && traceIfFalse "liabilities < 0" (liabilities bs cr >= zero)
-    && traceIfFalse "equity < 0" (equity bs cr >= zero)
+isValidState sc bs cr = isRight (checkValidState sc bs cr)
+
+{-# INLINEABLE checkValidState #-}
+checkValidState :: Stablecoin -> BankState -> ConversionRate -> Either InvalidStateReason ()
+checkValidState sc bs@BankState{bsReservecoins, bsReserves, bsStablecoins} cr = do
+    -- TODO: Do we need a validation type in the state machine lib?
+    unless (bsReservecoins >= RC 0) (Left NegativeReserveCoins)
+    unless (bsReserves >= BC 0) (Left NegativeReserves)
+    unless (bsStablecoins >= SC 0) (Left NegativeStablecoins)
+    unless (liabilities bs cr >= zero) (Left NegativeLiabilities)
+    unless (equity bs cr >= zero) (Left NegativeEquity)
+
+    let actualReserves = fmap fromInteger bsReserves
+        allowedReserves = (,) <$> minReserve sc cr bs <*> maxReserve sc cr bs
```

Comment:
Worth a comment that the `Just + Nothing`/`Nothing + Just` cases can't occur, since they both return `Nothing` iff the other returns `Nothing`.